### PR TITLE
Functional interface annotation added when needed

### PR DESCRIPTION
**Pull request by rpau**
Hi,

I am sending to you an automatic correction to add the @FunctionalInterface annotation in interfaces that only contains one method. If it is too much to verify, let me know how would you like to proceed. For example: per module. 

According to the nemo project, it should reduce 2 hours approximately of technical debt :)

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/qualitygate/ProjectQgateAssociationMapper.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import java.util.List;
 import org.apache.ibatis.annotations.Param;
 
+@FunctionalInterface
```
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/qualitygate/ProjectQgateAssociationMapper.java]: MyBatis mappers do not aim to FunctionalInterfaces.

File: sonar-plugin-api/src/main/java/org/sonar/api/component/RubyComponentService.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
 /**
  * @since 3.6
  */
+@FunctionalInterface
 @ServerSide
 public interface RubyComponentService {
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/component/RubyComponentService.java]: This class does not aim to be a FunctionalInterface.

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/batch/IssueFilter.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
  * @since 4.0
  * @deprecated since 5.3. Use {@link org.sonar.api.scan.issue.filter.IssueFilter} instead.
  */
+@FunctionalInterface
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/batch/IssueFilter.java]: Deprecated classes should not be touched.

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/MeasureBuilder.java [hunk 4]
```diff
@@ -27,6 +27,7 @@
  * @since 1.10
  * @deprecated since 5.6. To create a new measure on scanner side use {@link SensorContext#newMeasure()}
  */
+@FunctionalInterface
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/MeasureBuilder.java]: Builders are not FunctionalInterface.

**Comment by simonbrandhof** [comment]
Thanks for the contribution but @FunctionalInterface should not be put on all single-method interfaces. It depends on the responsibility of the interface. Moreover deprecated classes are usually not touched.

**Comment by rpau** [comment]
I have removed from the list those classes that you reported me that are not functional interfaces

**Comment by simonbrandhof** [comment]
Thanks, but as there are many touched classes that was only a partial review :-)

**Comment by rpau** [comment]
So, would you prefer I split them by modules?

**Comment by simonbrandhof** [comment]
No. You should add the annotation only where it makes sense, not on all Single Abstract Method (SAM) interfaces. For example ignore data structures and mybatis mappers.

**Comment by simonbrandhof** [comment]
Thanks for the contribution but we will fix add the annotation on missing places during future refactorings.

**Pull request closed without merging**